src/state.rs

```diff
@@ -365,78 +370,126 @@ impl TabExpandedString {
             if new < self.prev.0 {
-                self.reset(now);
+                self.reset();
             }
             return;
         }
 
-        let elapsed = now - self.prev.1;
-        let divisor = delta as f64;
-        let mut batch = 0.0;
-        if divisor != 0.0 {
-            batch = duration_to_secs(elapsed) / divisor;
-        };
+        let delta_steps = new - self.prev.0;
+        let delta_t = duration_to_secs(now - self.prev.1);
+
+        // the rate of steps we saw in this tick
+        let new_steps_per_second = delta_steps as f64 / delta_t;
 
-        self.steps[self.pos as usize] = batch;
-        self.pos = (self.pos + 1) % 16;
-        if !self.full && self.pos == 0 {
-            self.full = true;
+        // update the estimate
+        let weight = self.weight(delta_t);
+        self.smoothed_steps_per_sec =
+            self.smoothed_steps_per_sec * weight + new_steps_per_second * (1.0 - weight);
+
+        if self.double_smoothing {
+            // correct for bias in initial Estimator value (see steps_per_second comment)
+            let delta_t_start = duration_to_secs(now - self.start);
+            let debias = 1.0 - self.weight(delta_t_start);
+
+            self.double_smoothed_steps_per_sec = self.double_smoothed_steps_per_sec * weight
+                + self.smoothed_steps_per_sec * (1.0 - weight) / debias;
         }
 
         self.prev = (new, now);
     }
 
-    pub(crate) fn reset(&mut self, now: Instant) {
-        self.pos = 0;
-        self.full = false;
+    pub(crate) fn reset(&mut self) {
+        let now = Instant::now();
+        self.smoothed_steps_per_sec = 0.0;
+        self.double_smoothed_steps_per_sec = 0.0;
         self.prev = (0, now);
+        self.start = now;
     }
 
-    /// Average time per step in seconds, using rolling buffer of last 15 steps
-    fn seconds_per_step(&self) -> f64 {
-        let len = self.len();
-        self.steps[0..len].iter().sum::<f64>() / len as f64
+    /// Get the appropriate dilution weight for Estimator data given an age (in seconds)
+    fn weight(&self, age: f64) -> f64 {
+        // The goal is for data to reach a dilution of 0.1 at age `exponent_weight_seconds`.
+        // We achieve this by diluting with the equation `weight ^ (ews / delta_t) = 0.1`.
+        10_f64.powf(-age / self.exponent_weight_seconds)
     }
 
-    fn len(&self) -> usize {
-        match self.full {
-            true => 16,
-            false => self.pos as usize,
+    /// Average time per step in seconds, using single or double exponential smoothing
+    fn steps_per_second(&self) -> f64 {
+        let now = Instant::now();
+
+        // reweight to account for the passage of time since last tick
+        let delta_t = duration_to_secs(now - self.prev.1);
+        let reweight = self.weight(delta_t);
+
+        // The Estimator is initialized with zero values, which are included in the underlying
```

Comment:
See my other comment about avoiding the pathologies of an uninitialized estimate.